diff/diff.go

```diff
@@ -0,0 +1,534 @@
+		case 0:
+			// update
+			deleteSQL := generateDML("delete", rowsData2[index2], orderKeyCols, table.Info, table.Schema)
+			replaceSQL := generateDML("replace", rowsData1[index1], orderKeyCols, table.Info, table.Schema)
+			log.Infof("[update] delete: %s,\n replace: %s", deleteSQL, replaceSQL)
+			df.wg.Add(2)
+			df.sqlCh <- deleteSQL
+			df.sqlCh <- replaceSQL
+			index1++
+			index2++
+		}
+	}
+
+	return equal, nil
+}
+
+// WriteSqls write sqls to file
+func (df *Diff) WriteSqls() {
+	defer df.wg.Done()
+	for {
+		select {
+		case dml, ok := <-df.sqlCh:
+			if !ok || dml == "end" {
+				return
+			}
+
+			_, err := df.fixSQLFile.WriteString(fmt.Sprintf("%s\n", dml))
+			if err != nil {
+				log.Errorf("write sql: %s failed, error: %v", dml, err)
+			}
+			df.wg.Done()
+		default:
+		}
+	}
+}
+
+func generateDML(tp string, data map[string][]byte, keys []*model.ColumnInfo, table *model.TableInfo, dbName string) (sql string) {
+	// TODO: can't distinguish NULL between ""
+	switch tp {
+	case "replace":
+		colNames := make([]string, 0, len(table.Columns))
+		values := make([]string, 0, len(table.Columns))
+		for _, col := range table.Columns {
+			colNames = append(colNames, col.Name.O)
+			if needQuotes(col.FieldType) {
+				values = append(values, fmt.Sprintf("\"%s\"", string(data[col.Name.O])))
+			} else {
+				values = append(values, string(data[col.Name.O]))
+			}
+		}
+
+		sql = fmt.Sprintf("REPLACE INTO `%s`.`%s`(%s) VALUES (%s);", dbName, table.Name, strings.Join(colNames, ","), strings.Join(values, ","))
+	case "delete":
+		kvs := make([]string, 0, len(keys))
+		for _, col := range keys {
+			if needQuotes(col.FieldType) {
+				kvs = append(kvs, fmt.Sprintf("%s = \"%s\"", col.Name.O, string(data[col.Name.O])))
+			} else {
+				kvs = append(kvs, fmt.Sprintf("%s = %s", col.Name.O, string(data[col.Name.O])))
+			}
+		}
+		sql = fmt.Sprintf("DELETE FROM `%s`.`%s` where %s;", dbName, table.Name, strings.Join(kvs, " AND "))
+	default:
+		log.Errorf("unknow sql type %s", tp)
+	}
+
+	return
+}
+
+func needQuotes(ft types.FieldType) bool {
+	return !(pkgdb.IsNumberType(ft.Tp) || pkgdb.IsFloatType(ft.Tp))
+}
+
+func compareData(map1 map[string][]byte, map2 map[string][]byte, orderKeyCols []*model.ColumnInfo) (bool, int32, error) {
+	var (
+		equal        = true
+		data1, data2 []byte
+		key          string
+		ok           bool
+		cmp          int32
+	)
+
+	for key, data1 = range map1 {
+		if data2, ok = map2[key]; !ok {
+			return false, 0, errors.Errorf("don't have key %s", key)
+		}
+		if string(data1) == string(data2) {
+			continue
+		}
+		equal = false
+		log.Errorf("find difference data, data1: %s, data2: %s", map1, map2)
+		break
+	}
+	if equal {
+		return true, 0, nil
+	}
+
+	for _, col := range orderKeyCols {
+		if data1, ok = map1[col.Name.O]; !ok {
+			return false, 0, errors.Errorf("don't have key %s", col.Name.O)
+		}
+		if data2, ok = map2[col.Name.O]; !ok {
+			return false, 0, errors.Errorf("don't have key %s", col.Name.O)
```

Comment:
differ `map1` / `map2` to assist deubg?

Reply:
what are you mean?

Reply:
`return false, 0, errors.Errorf("map1/map2 don't have key %s", col.Name.O)` to distinguish with each other?